<a href="https://colab.research.google.com/github/juan-lawrie/Repaso-Ciencias-de-datos/blob/main/Copia_de_04_%5BPr%C3%A1ctica%5D_Listas%2C_indexaci%C3%B3n_y_mutabilidad_Iteraci%C3%B3n_y_ciclo_while.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 📌 Tema 4 Práctica: Listas, indexación y mutabilidad. Iteración y ciclo while


❗Antes de comenzar con la clase, hacete una copia de este documento así podés editarlo y guardar los cambios.

Hacé clic acá arriba, donde dice "Copiar en Drive".

Luego seguís trabajando en esa copia y te queda todo guardado en tu Drive.

***Se realizará una puesta en común en clase de los ejercicios 2, 5, 6, 8, 9 y 10.*** 🚀

---

## 📝 Ejercicio 1

En una ciudad se registró la temperatura media (en grados) durante un año, de enero a diciembre:

Los datos son los siguientes: 22, 24, 27, 30, 33, 35, 36, 35, 32, 28, 24, 22.

Guardá los datos de temperatura en una lista:

In [ ]:
temperaturas =[22, 24, 27, 30, 33, 35, 36, 35, 32, 28, 24, 22]
print("Temperaturas:", temperaturas)

Temperaturas: [22, 24, 27, 30, 33, 35, 36, 35, 32, 28, 24, 22]


Mostrá las temperaturas de enero, junio y diciembre:

In [ ]:
print(f"Temperatura de enero: {temperaturas[0]}°")
print(f"Temperatura junio: {temperaturas[5]}°")
print(f"Temperatura diciembre: {temperaturas[-1]}°")

Temperatura de enero: 22°
Temperatura junio: 35°
Temperatura diciembre: 22°


Mostrá el otoño completo (marzo, abril y mayo) usando rebanado:

In [ ]:
print(f"Temperatura  de marzo: {temperaturas[2]}°")
print(f"Temperatura de abril: {temperaturas[3]}°")
print(f"Temperatura de mayo: {temperaturas[4]}°")

Temperatura  de marzo: 27°
Temperatura de abril: 30°
Temperatura de mayo: 33°


Reemplazá el valor de julio por 37:

In [ ]:
temperaturas[6] = 37
print("Temperaturas luego de la modificación:", temperaturas)

Temperaturas luego de la modificación: [22, 24, 27, 30, 33, 35, 37, 35, 32, 28, 24, 22]


Mostrá la primera mitad del año:

In [ ]:
print(temperaturas[0:6])

[22, 24, 27, 30, 33, 35]


Mostrá la segunda mitad del año usando índices negativos:

In [ ]:
print(temperaturas[-6:0])

[22, 24, 27]


## 📝 Ejercicio 2

En una encuesta realizada a 10 personas se les pidió que calificaran un servicio con puntajes del 1 al 5.

Los datos son los siguientes: 5, 4, 3, 5, 2, 1, 4, 5, 3, 2

Guardá las respuestas en una lista:

In [ ]:
respuestas =  [5, 4, 3, 5, 2, 1, 4, 5, 3, 2]
print("Respuestas:",respuestas )

Respuestas: [5, 4, 3, 5, 2, 1, 4, 5, 3, 2]


Mostrá la primera y última respuesta:

In [ ]:
print(respuestas[0])
print(respuestas[-1])

5
2


Mostrá las 4 respuestas centrales:

In [ ]:
print(respuestas[3:7])

(5, 2, 1, 4)


Intercambiá las respuestas de la persona 2 y la persona 4:

In [ ]:
#aux = respuestas[]
respuestas[1] = 5
respuestas[3] = 4

print("Respuestas luego del intercambio:", respuestas)

Respuestas luego del intercambio: [5, 5, 3, 4, 2, 1, 4, 5, 3, 2]


Mostrá solo las respuestas de personas en posiciones impares (1, 3, 5...).

*Pista*: podés usar rebanado con un tercer índice para definir un paso o salto: `inicio:fin:paso`

In [ ]:
print(respuestas[])

Mostrá solo las respuestas en posiciones pares (0, 2, 4...):

In [ ]:
print(respuestas[])

## 📝 Ejercicio 3

Una cafetería registró los pedidos diarios durante 2 semanas:

In [ ]:
pedidos_semana_1 = [120, 135, 150, 145, 160, 180, 200]  # Lunes a domingo
pedidos_semana_2 = [130, 140, 155, 150, 170, 190, 210]

Uní ambas semanas en una lista completa de 14 días:

In [ ]:
pedidos_totales = pedidos_semana_1 + pedidos_semana_2
#pedidos_totales.insert(-1,pedidos_semana_2)
print("Pedidos totales:", pedidos_totales)

Pedidos totales: [120, 135, 150, 145, 160, 180, 200, 130, 140, 155, 150, 170, 190, 210]


Extraé los pedidos del primer fin de semana (sábado y domingo de la semana 1) y luego calculá y mostrá la suma total de pedidos:

In [ ]:
fin_de_semana_1 = pedidos_totales[5:7] #Preguntar
total_fin_de_semana_1 = fin_de_semana_1[0] + fin_de_semana_1[1]
print("Pedidos totales fin de semana 1:", total_fin_de_semana_1 )

Pedidos totales fin de semana 1: 380


Mostrá los pedidos de la primera y segunda semana por separado:

In [ ]:
primera_mitad = pedidos_totales[0:7]
segunda_mitad = pedidos_totales[7:14]

print("Pedidos semana 1:", primera_mitad)
print("Pedidos semana 2:", segunda_mitad)

Pedidos semana 1: [120, 135, 150, 145, 160, 180, 200]
Pedidos semana 2: [130, 140, 155, 150, 170, 190, 210]


Reemplazá el valor del primer lunes (semana 1) por 125:

In [ ]:
pedidos_totales[0] = 125
print("Pedidos totales luego del reemplazo:",pedidos_totales )

Pedidos totales luego del reemplazo: [125, 135, 150, 145, 160, 180, 200, 130, 140, 155, 150, 170, 190, 210]


Mostrá los días con 150 o más pedidos por semana (usá rebanado y concatenación para elegir intervalos y unir):

In [ ]:
altos_semana_1 = [pedidos_totales[2]] + pedidos_totales[4:7]
altos_semana_2 = pedidos_totales[-5:]

print("Pedidos altos semana 1:", altos_semana_1)
print("Pedidos altos semana 2:", altos_semana_2)

Pedidos altos semana 1: [150, 160, 180, 200]
Pedidos altos semana 2: [155, 150, 170, 190, 210]


## 📝 Ejercicio 4

Se registraron las ventas de 3 productos en 4 semanas:

In [ ]:
producto_A = [10, 12, 13, 11]
producto_B = [20, 18, 17, 21]
producto_C = [15, 14, 16, 15]

Calculá y mostrá la venta total de la primera semana (sumá manualmente con índices):

In [ ]:
total_semana_1 = producto_A[0] + producto_A[1] + producto_A[2] + producto_A[3]
print("Total semana 1:", total_semana_1 )

Total semana 1: 46


Mostrá la última venta de cada producto:

In [ ]:
print("Última venta producto A:", producto_A[-1] )
print("Última venta producto B:", producto_B[-1])
print("Última venta producto C:", producto_C[-1])

Última venta producto A: 11
Última venta producto B: 21
Última venta producto C: 15


Armá una lista con todas las ventas de los 3 productos usando concatenación:

In [ ]:
ventas_totales = producto_B + producto_A + producto_C
print("Ventas totales:",ventas_totales )

Ventas totales: [20, 18, 17, 21, 10, 12, 13, 11, 15, 14, 16, 15]


Mostrá solo las ventas de producto B usando la lista conjunta:

In [ ]:
print("Ventas producto B:",producto_B[0:5] )

Ventas producto B: [20, 18, 17, 21]


Reemplazá los dos últimos valores del producto C por 15 en la lista conjunta usando rebanado y repetición:

In [ ]:
#[-4:-1:2]
ventas_totales[-4:-1:2] = [15] * 2
print("Ventas totales luego de la modificación:", ventas_totales)

Ventas totales luego de la modificación: [20, 18, 17, 21, 10, 12, 13, 11, 15, 14, 15, 15]


## 📝 Ejercicio 5

En un consultorio de una clínica se registraron los turnos del día para la sección de **cardiología**.

Los datos son los siguientes: Ana, Luis, María, Jorge, Sofía

Guardá los turnos en una lista:

In [ ]:
turnos_cardiologia = ['Ana', 'Luis', 'María', 'Jorge', 'Sofía']
print("Turnos cardiología:", turnos_cardiologia)

Turnos cardiología: ['Ana', 'Luis', 'María', 'Jorge', 'Sofía']


Mostrá el primer y último turno del día:

In [ ]:
print("Primer turno:", turnos_cardiologia[0])
print("Último turno:", turnos_cardiologia[-1])

Primer turno: Ana
Último turno: Sofía


Modificá el segundo turno por "Lucía":

In [ ]:
turnos_cardiologia[1] = "Lucia"
print("Turnos actualizados:",turnos_cardiologia )

Turnos actualizados: ['Ana', 'Lucia', 'María', 'Jorge', 'Sofía']


Mostrá los turnos centrales (segundo, tercer y cuarto turno):

In [ ]:
print("Turnos centrales:",turnos_cardiologia[1:4] )

Turnos centrales: ['Lucia', 'María', 'Jorge']


Agregá dos turnos más al final: "Pedro" y "Elena" (usando concatenación):

In [ ]:
turnos_cardiologia = turnos_cardiologia + ["Pedro","Elena"]
print("Turnos con agregados:", turnos_cardiologia)

Turnos con agregados: ['Ana', 'Lucia', 'María', 'Jorge', 'Sofía', 'Pedro', 'Elena']


Eliminá los dos primeros turnos (usando rebanado):

In [ ]:
turnos_cardiologia = turnos_cardiologia[2:]
print("Turnos luego de eliminar los dos primeros:", turnos_cardiologia)

Turnos luego de eliminar los dos primeros: ['María', 'Jorge', 'Sofía', 'Pedro', 'Elena']


Ahora supongamos que en el **laboratorio** también se registraron los turnos para otra sección de atención.

Observá este código:

In [ ]:
turnos_laboratorio = ["Carlos", "Rocío", "Andrés", "Valeria", "Noelia"]

turnos_lab_A = turnos_laboratorio
turnos_lab_B = turnos_laboratorio[:]

Modificá el primer turno de la lista `turnos_lab_A` por "Gabriel":

In [ ]:
turnos_lab_A[0] = "Gabriel"
print(turnos_lab_A )

['Gabriel', 'Rocío', 'Andrés', 'Valeria', 'Noelia']


Mostrá todas las listas:

In [ ]:
print("Turnos laboratorio (original):",turnos_laboratorio )
print("Turnos en laboratorio A:", turnos_lab_A )
print("Turnos en laboratorio B:", turnos_lab_B )

Turnos laboratorio (original): ['Gabriel', 'Rocío', 'Andrés', 'Valeria', 'Noelia']
Turnos en laboratorio A: ['Gabriel', 'Rocío', 'Andrés', 'Valeria', 'Noelia']
Turnos en laboratorio B: ['Carlos', 'Rocío', 'Andrés', 'Valeria', 'Noelia']


📌 Preguntas:
1. ¿Qué diferencias observás entre las listas `turnos_lab_A` y `turnos_lab_B`?
2. ¿Qué pasó con la lista `turnos_laboratorio` después de modificar `turnos_lab_A`?
3. ¿Por qué `turnos_lab_B` no se modificó?
4. ¿Qué podés decir sobre la diferencia entre asignar una lista y copiarla con `[:]`?

✏️ Respuestas:

1. ...
2. ...
3. ...
4. ...

## 📝 Ejercicio 6

Retomemos los datos de ventas de la tienda que vimos en la clase 4.

Los datos (en miles de unidades) para cada mes son: 320, 280, 310, 290, 340, 360, 410, 390, 370, 350, 300, 330.

Completá el código para calcular el total de ventas:

In [ ]:

ventas = [320, 280, 310, 290, 340, 360, 410, 390, 370, 350, 300, 330]

i = 0


suma_ventas = 320 + 280 + 310 + 290 + 340 + 360 + 410 + 390 + 370 + 350 + 300 + 330
n = len(ventas)
print(n)
while i>=len(ventas):

    i= i +1


    print("Suma total de ventas:", suma_ventas)
    print(i)


12


📌 Preguntas:

1. ¿Qué valor devuelve `len(ventas)`?
2. ¿Cuántas iteraciones hace el ciclo antes de terminar?
3. ¿Cuál es el valor de `i` y de `suma_ventas` al finalizar la iteración 3?
4. ¿Qué pasaría si usáramos `i <= len(notas)` en la condición del `while`?

✏️ Respuestas:

1. 12
2. 12
3. Al finalizar el ciclo i = 12 y suma_ventas = 4050
4. NameError: name 'notas' is not defined

Calculá la media de ventas:

In [ ]:
media_ventas = (320+280+310+290+340+360+410+390+370+350+300+330)/12
print("Media de ventas:", media_ventas)

Media de ventas: 337.5


Mostrá en pantalla las ventas de los primeros 6 meses usando un ciclo `while`:

In [ ]:

while True:
    print(ventas[0:6])
    break

[320, 280, 310, 290, 340, 360]


Contá cuántos meses tuvieron ventas mayores a 350:

In [ ]:
ventas = [320, 280, 310, 290, 340, 360, 410, 390, 370, 350, 300, 330]





def filtrar_en_ventas(ventas):
    return 350 < ventas

filtrar_ventas = filter(filtrar_en_ventas,ventas)

lista_ventas_filtrada = list(filtrar_ventas)
cantidad_de_meses_con_ventas_mayores_a_350 = len(lista_ventas_filtrada)

print(lista_ventas_filtrada)
print(cantidad_de_meses_con_ventas_mayores_a_350)

[360, 410, 390, 370]
4


Contá cuántos meses tuvieron ventas menores o iguales a 300:

In [ ]:
def filtrar_en_ventas(ventas):
    return 300 >= ventas

filtrar_ventas = filter(filtrar_en_ventas,ventas)

lista_ventas_filtrada = list(filtrar_ventas)
cantidad_de_meses_con_ventas_mayores_a_300 = len(lista_ventas_filtrada)

print(lista_ventas_filtrada)
print(cantidad_de_meses_con_ventas_mayores_a_300)

[280, 290, 300]
3


Mostrá los índices de los meses en los que las ventas fueron mayores a 320 y menores a 370:

In [ ]:
ventas.sort()
print(ventas)
while True:
  print(ventas[5:9])
  break


[280, 290, 300, 310, 320, 330, 340, 350, 360, 370, 390, 410]
[330, 340, 350, 360]


Calculá la media de ventas solo de los meses que tuvieron más de 350.

*Pista*: vas a necesitar un acumulador para la suma y un contador para saber cuántos fueron.

In [ ]:

def filtrar_en_ventas(ventas):
    return 350 < ventas

filtrar_ventas = filter(filtrar_en_ventas,ventas)

lista_ventas_filtrada = list(filtrar_ventas)
cantidad_de_meses_con_ventas_mayores_a_350 = len(lista_ventas_filtrada)



media_ventas_meses_con_mas_de_350 = (360 + 410 + 390 + 370)/4


print(lista_ventas_filtrada)


print("Meses con ventas superiores a 350:", cantidad_de_meses_con_ventas_mayores_a_350)
print("Media de ventas superiores a 350:", media_ventas_meses_con_mas_de_350)

[360, 370, 390, 410]
Meses con ventas superiores a 350: 4
Media de ventas superiores a 350: 382.5


Ahora queremos saber cuántos meses de ventas se necesitan sumar desde el principio hasta que el total acumulado supere 900.

Para eso, usá un ciclo `while` y una **variable bandera** que se vuelva `True` cuando se supere ese total:

In [ ]:
suma_acumulada = 0
supero_limite = False

i = 0
while
    suma_acumulada =
    if
        supero_limite =
    i =

print("Suma acumulada:", suma_acumulada)
print("Meses necesarios para superar 900:", i)

SyntaxError: invalid syntax (4176507295.py, line 1)

## 📝 Ejercicio 7

En una materia se dispone de las notas finales de un examen y se quiere calcular el desvío estándar.

Sin embargo, el código que tenemos tiene errores. Corregilo para que funcione bien.

In [ ]:
notas = [7, 8, 9, 6, 10, 5]

i = 0
suma_notas = 1

while i < len(notas):
  suma_notas = suma_notas + notas[i]
  i = i

media_notas = suma_nota / len(notas)

i = 0

while i < len(notas):
    dif_notas = notas[i] - media_notas
    suma_dif_notas = suma_dif_notas + dif_notas**2
    i = i+1

desvio_notas = (suma_dif_notas / len(notas)) ** 0.5
print("El desvío estándar es:", desvio_notas)

KeyboardInterrupt: 

📌 Preguntas:

1. ¿Por qué usamos dos ciclos `while`?
2. ¿Qué clase de variables son `suma_notas` y `suma_dif_notas`?

✏️ Respuestas:

1. ...
2. ...

Modificá el código para que `len(notas)` se use una sola vez:

In [ ]:
# Escribí tu código acá

¿Cómo calcularías el rango (máximo-mínimo) usando los índices de la lista? Probalo:

In [ ]:
# Escribí tu código acá

## 📝 Ejercicio 8

Observá el código en la celda a continuación:

In [ ]:
nombres = ["Luna", "Oscar", "Mateo", "Julia", "Ana"]
asistencias = [True, False, True, False, True]

n = len(nombres)
presentes = []

i = 0
while i < n:
    if asistencias[i] == True:
        presentes = presentes + [nombres[i]]
    i = i + 1

print(presentes)

['Luna', 'Mateo', 'Ana']


📌 Preguntas:

1. Explicá con tus palabras lo que hace el código.
2. ¿Qué tipo de datos contienen las listas `nombres` y `asistencias`?
3. ¿Qué se guarda en la variable `n`? ¿Sería lo mismo si se usara `len(asistencias)` en este caso?
4. ¿Que hace la línea  `presentes = []`?
5. ¿Qué operación se hace en `presentes = presentes + [nombres[i]]`?

✏️ Respuestas:

1. Te muestra las personas presentes, osea que en la lista de asistencias estan marcada con el respectivo indice que le corresponde como True.
2. La lista de nombres es una lista de strings y la de asistencias es una lista de booleanos
3. La variable n guarda la longitud de la lista nombres
4. La linea presentes = [] crea una lista vacia.
5. La operacion presentes = presentes + [nombres[i]] iguala la lista vacia presentes con presentes y despues la concatena con la lista nombre pasando le el indice i a la lista nombre que se va a ir sumando en cada ciclo del while mientras que el i sea menor que el n.

## 📝 Ejercicio 9

Una organización está analizando los datos de edad de 15 participantes de un taller. Las edades recolectadas son:

In [ ]:
edades = [28, 35, 42, 51, 63, 40, 47, 54, 76, 33, 45, 60, 29, 41, 38]

Se desea clasificar a cada persona según su grupo etario:
- 🟢 **Adulto joven**: de 20 a 39 años inclusive.
- 🔵 **Adulto**: de 40 a 59 años inclusive.
- 🟣 **Persona mayor**: 60 años o más.

Completá el código para armar una lista que contenga el grupo etario según la edad:

In [ ]:
# Definí una lista vacía para guardar los grupos etarios
grupos = []

# Recorré la lista de edades y asigná el grupo etario correspondiente
i = 0
while i < len():
    if 20 <= edades[i] and edades[i] <= 39:
        grupos = grupos + ["Adulto joven"]
    elif

    i = i + 1

print("Grupos etarios:", grupos)

Si en lugar inicializar `grupos` con una lista vacía, hiciéramos  `grupos = [0] * len(edades)`.

¿Qué necesitás modificar dentro del ciclo `while` para que funcione bien?

In [ ]:
# Escribí tu código acá

Calculá las **frecuencias absolutas y relativas** para cada categoría de grupo:

In [ ]:
# Frecuencias absolutas

conteo_adulto_joven =
conteo_adulto =
conteo_persona_mayor

i =
while
    if grupos[i] == "Adulto joven":
        conteo_adulto_joven
    elif

    i =

print("Frecuencia absoluta:")
print("Adulto joven:", )
print("Adulto:", )
print("Persona mayor:", )

In [ ]:
# Frecuencias relativas

total =

proporcion_adulto_joven =
proporcion_adulto
proporcion_persona_mayor

print("Frecuencia relativa:")
print("Adulto joven:", )
print("Adulto:", )
print("Persona mayor:", )

📌 Preguntas:
1. ¿Qué tipo de dato estadístico son *edades* y *grupos*?
2. ¿Cuál es el grupo etario mayoritario? ¿Y el minoritario?

✏️ Respuestas:

1. ...
2. ...

Con los datos de grupos y edades, calculá la **media de edad dentro de cada grupo etario**:

In [ ]:
suma_edades_adulto_joven = 0
suma_edades_adulto
suma_edades_persona_mayor

i =
while
    if
    elif
    else
    i =

media_edades_adulto_joven =
media_edades_adulto
media_edades_persona_mayor

print("Media de edades Adulto joven:", )
print("Media de edades Adulto:", )
print("Media de edades Persona mayor:", )

## 📝 Ejercicio 10

Una nutricionista está analizando las alturas de 12 personas que participaron en una consulta grupal.

Los datos disponibles son:

In [ ]:
alturas_cm = [152, 170, 165, 180, 155, 160, 175, 168, 182, 158, 162, 177]


Las alturas fueron recolectadas en centímetros, pero para los informes debe trabajarlas en **metros**.

Escribí un código para transformar las alturas a metros y guardarlas en una nueva lista:

In [ ]:
# Escribí tu código acá

resultado_dividir = datos/100
print(list(resultado_dividir))


[np.float64(1.52), np.float64(1.7), np.float64(1.65), np.float64(1.8), np.float64(1.55), np.float64(1.6), np.float64(1.75), np.float64(1.68), np.float64(1.82), np.float64(1.58), np.float64(1.62), np.float64(1.77)]


Ahora calculá la media y el desvío estándar de las alturas en metros:

In [ ]:
# Escribí tu código acá

📌 Preguntas:
1. ¿Qué tipo de dato estadístico es la *altura*?
2. ¿En qué unidades están la media y el desvío estándar?
3. ¿Qué significa que el desvío estándar sea alto o bajo en este contexto?
4. Si otro grupo tiene alturas con una media de 1.68 m y un desvío de 0.02, ¿qué podría concluir la nutricionista respecto a ese grupo?

✏️ Respuestas:
1. ...
2. ...
3. ...
4. ...